Immutable roles (#1984)
neo4j/docs-operations: modules/ROOT/pages/authentication-authorization/privileges-and-roles-immutable.adoc

[role=label--new-5.24]
[role=enterprise-edition not-on-aura]
[[access-control-privileges-immutable]]
= Immutable privileges and roles
:description: This section explains how to use Cypher to manage immutable privileges and roles.

[NOTE]
====
Immutable roles are newly introduced in Neo4j 5.24. Immutable privileges have been available since Neo4j 5.0.
====

[[access-control-privileges-immutable-admin]]
== How to administer immutable privileges and roles

> Unlike for regular privileges and roles, immutable privileges and roles cannot be administered by users with xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-privilege-management[PRIVILEGE MANAGEMENT] and xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-role-management[ROLE MANAGEMENT] privileges. Instead, they can only be administered when auth is disabled -- that is, when the configuration setting <<config_dbms.security.auth_enabled,`dbms.security.auth_enabled`>> is set to `false`.

Reviewer: ```suggestion
Unlike regular privileges and roles, immutable privileges and roles cannot be administered by users with xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-privilege-management[PRIVILEGE MANAGEMENT] and xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-role-management[ROLE MANAGEMENT] privileges. 
Instead, they can only be administered when auth is disabled -- that is, when the configuration setting <<config_dbms.security.auth_enabled,`dbms.security.auth_enabled`>> is set to `false`.
```

[CAUTION]

> It is intended that administering immutable privileges and roles is an extraordinarily rare scenario and one which should be undertaken with extreme caution (i.e. when the dbms has been isolated by some other means and unauthorized access can be reliably prevented). It is considered to be the kind of action which may be performed once during the commissioning phase of a database.

Reviewer: This is a bit dramatic
```suggestion
Immutable privileges and roles should only be used in a scenario where changes are rare. 
They are intentionally difficult to change and thus changes should be undertaken with caution (i.e. when the DBMS has been isolated by some other means and unauthorized access can be reliably prevented).
It is considered to be the kind of action which may be performed once during the commissioning phase of a DBMS.
```

> [role=label--new-5.24]

Author: ```suggestion
[role=label--new-5.26]
```

Reviewer 2: Yes, I've already addressed those in my local branch.

[role=enterprise-edition not-on-aura]
[[access-control-privileges-immutable]]
= Immutable privileges and roles
:description: This section explains how to use Cypher to manage immutable privileges and roles.

[NOTE]
====

> Immutable roles were introduced in Neo4j 5.24. Immutable privileges have been available since Neo4j 5.0.

Author: ```suggestion
Immutable roles were introduced in Neo4j 5.26. Immutable privileges have been available since Neo4j 5.0.
```

(see xref:authentication-authorization/privileges-and-roles-immutable.adoc#access-control-privileges-immutable-usecase[here] for examples).
. Change the config setting xref:configuration/configuration-settings.adoc#config_dbms.security.auth_enabled[`dbms.security.auth_enabled`] back to `true`.
. Restart the DBMS.

Privileges and roles created in this way will now appear to all users as an immutable part of the DBMS.
The only way to subsequently change or remove them it would be to repeat the process of setting xref:configuration/configuration-settings.adoc#config_dbms.security.auth_enabled[`dbms.security.auth_enabled`] to `false`.

[[access-control-privileges-immutable-usecase]]
== When to use immutable privileges and roles

Immutable privileges and roles are useful for restricting the actions of users who can themselves xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-privilege-management[administer privileges] and xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-role-management[roles].

=== Immutable privileges

> You may want to prevent all users from performing Database Management, even the `admin` user (who are themselves able to add or remove privileges).

Reviewer 2: I find this section overly explanatory. Instead of telling them how to prevent users from performing database management, we first tell them what not to do and then give them an example of using immutable privileges. Would it be more helpful if you directly told them how to do it?

Author: yes good point. Will re-phrase

Reviewer 2: I'll do it as part of my review commit —no problem. I just wondered if there is any particular reason for explaining to the users the wrong way of doing it before telling them how to do it.
